validator/delphi_validator/validate.py

```diff
@@ -0,0 +1,742 @@
+        # difference, 2 * mean(differences) / sum(means) of two groups.
+        df_all = pd.melt(
+            df_all, id_vars=["geo_id", "type"], value_vars=["val", "se", "sample_size"]
+        ).pivot(
+            index=("geo_id", "variable"), columns="type", values="value"
+        ).reset_index(
+            ("geo_id", "variable")
+        ).dropna(
+        ).assign(
+            type_diff=lambda x: x["test"] - x["reference"],
+            abs_type_diff=lambda x: abs(x["type_diff"])
+        ).groupby(
+            "variable", as_index=False
+        ).agg(
+            mean_type_diff=("type_diff", "mean"),
+            mean_abs_type_diff=("abs_type_diff", "mean"),
+            mean_test_var=("test", "mean"),
+            mean_ref_var=("reference", "mean")
+        ).assign(
+            mean_stddiff=lambda x: 2 *
+            x["mean_type_diff"] / (x["mean_test_var"] + x["mean_ref_var"]),
+            mean_stdabsdiff=lambda x: 2 *
+            x["mean_abs_type_diff"] / (x["mean_test_var"] + x["mean_ref_var"])
+        )[["variable", "mean_stddiff", "mean_stdabsdiff"]]
+
+        # Set thresholds for raw and smoothed variables.
+        classes = ['mean_stddiff', 'val_mean_stddiff', 'mean_stdabsdiff']
+        raw_thresholds = pd.DataFrame([1.50, 1.30, 1.80], classes).T
+        smoothed_thresholds = raw_thresholds.apply(
+            lambda x: x/(math.sqrt(7) * 1.5))
+
+        switcher = {
+            'raw': raw_thresholds,
+            'smoothed': smoothed_thresholds,
+        }
+
+        # Get the selected thresholds from switcher dictionary
+        smooth_option = "smoothed" if sig in self.smoothed_signals else "raw"
+        thres = switcher.get(smooth_option, lambda: "Invalid smoothing option")
+
+        # Check if the calculated mean differences are high compared to the thresholds.
+        mean_stddiff_high = (abs(df_all["mean_stddiff"]) > thres["mean_stddiff"]).bool() or (
+            (df_all["variable"] == "val").bool() and (
+                abs(df_all["mean_stddiff"]) > thres["val_mean_stddiff"]).bool())
+        mean_stdabsdiff_high = (
+            df_all["mean_stdabsdiff"] > thres["mean_stdabsdiff"]).bool()
+
+        if mean_stddiff_high or mean_stdabsdiff_high:
+            self.raised_errors.append(ValidationError(
+                ("check_test_vs_reference_avg_changed",
+                 checking_date.date(), geo, sig),
+                (mean_stddiff_high, mean_stdabsdiff_high),
+                'Average differences in variables by geo_id between recent & reference data '
+                + '(either semirecent or from API) seem large --- either large increase '
+                + 'tending toward one direction or large mean absolute difference, relative '
+                + 'to average values of corresponding variables.  For the former check, '
+                + 'tolerances for `val` are more restrictive than those for other columns.'))
+
+    def validate(self, export_dir):
```

Comment:
It looks like this will only print output on failed checks (maybe I'm missing something though). Did you consider also printing out stats on success? It would be reassuring to see that N out N checks passed, for example, to confirm they are being run.

Reply:
That's right; currently, validation outputs the # of failed checks, the # of suppressed checks (would have failed, but they've been verified as false positives or otherwise not worth blocking data upload for), and messages from failed checks. It's a good idea to add the number of passed and total checks, as well.